tokio/src/sync/mutex.rs

```diff
@@ -86,11 +6,86 @@ use std::error::Error;
 use std::fmt;
 use std::ops::{Deref, DerefMut};
 
-/// An asynchronous mutual exclusion primitive useful for protecting shared data
+/// An asynchronous `Mutex`-like type.
 ///
-/// Each mutex has a type parameter (`T`) which represents the data that it is protecting. The data
-/// can only be accessed through the RAII guards returned from `lock`, which
-/// guarantees that the data is only ever accessed when the mutex is locked.
+/// This type acts similarly to an asynchronous [`std::sync::Mutex`], with one
+/// major difference: [`lock`] does not block.  
+/// Another difference is that the lock guard can be held across await points.
+/// This allows you to do something along the lines of:
```

Comment:
```suggestion
/// This type acts similarly to an asynchronous [`std::sync::Mutex`], with one
/// major difference: [`lock`] does not block. Another difference is that the
/// lock guard can be held across await points.
///
/// There are some situations where you should prefer the mutex from the
/// standard library. Generally this is the case if:
///
///  1. The lock does not need to be held across await points.
///  2. The duration of any single lock is near-instant.
///
/// On the other hand, the Tokio mutex is for the situation where the lock needs
/// to be held for longer periods of time, or across await points.
///
/// This allows you to do something along the lines of:
```
I'm not certain about the phrasing here, but I would really like to have a comment near the top that explains when to use which kind of lock.

Reply:
yeah makes total sense. I would just rephrase the subsequent phrase from `This allows you to do something along the lines of:` to `Example of using Mutex` or just `Example`
what do you think?

Reply:
Perhaps just an `# Examples` header.